### PR TITLE
feature-16

### PR DESCRIPTION
**Pull request by samuskitchen**
## Summary
- Added testing the controllers
- Added testing the models
## Trello Card
https://trello.com/c/eDUCaVcx/38-pull-request

### CODE REVIEW

File: src/test/java/wolox/training/models/BookTest.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                    bookRepository.save(oneTestBook);
+                });
+
+        assertEquals("Please check the author supplied, its null!.", thrown.getMessage());
+    }
+
+    @Test
+    public void whenFindByAuthor_thenReturnBook() {
+        Optional<Book> bookFound = bookRepository.findByAuthor("J. K. Rowling");
+        assertThat(oneTestBook.equals(bookFound.get())).isTrue();
```
Comment (Jimenarosello) [inline src/test/java/wolox/training/models/BookTest.java:77]: As an improvement, you could add a book that has a different author and verify that it is not returned by the method `findByAuthor` :)

File: src/test/java/wolox/training/models/UserTest.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+        assertEquals("Please check the username supplied, its null!.", thrown.getMessage());
+    }
+
+    @Test
+    public void whenFindByUsername_thenReturnUser() {
+        Optional<User> userFound = userRepository.findByUsername("SamusKitchen");
+        assertThat(oneTestUser.equals(userFound.get())).isTrue();
+        assertThat(oneTestUser.getBooks().size() == userFound.get().getBooks().size()).isTrue();
+        assertThat(oneTestBook.getAuthor().equals(userFound.get().getBooks().get(0).getAuthor())).isTrue();
```
Comment (Jimenarosello) [inline src/test/java/wolox/training/models/UserTest.java:87]: Same comment as before :)

**Review by Jimenarosello — COMMENTED** [review]
Nice job! I left a few commments. Correct me if I missed it but I think I didn't see tests for the cases that respond with a 404 status code, it would be good to add some tests for those scenarios 😊

File: src/test/java/wolox/training/controllers/BookControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+        oneTestBookCreated.setGenre("Fantasy");
+        oneTestBookCreated.setAuthor("John Ronald Reuel Tolkien");
+        oneTestBookCreated.setImage("https://images-na.ssl-images-amazon.com/images/I/81U5RVCuTHL.jpg");
+        oneTestBookCreated.setTitle("The Lord of the Rings");
+        oneTestBookCreated.setSubTitle("The Two Towers");
+        oneTestBookCreated.setPublisher("George Allen & Unwin");
+        oneTestBookCreated.setYear("1954");
+        oneTestBookCreated.setPages(352);
+        oneTestBookCreated.setIsbn("PR6039.O32 L6 1954, v.2");
```
Comment (JaimLus) [inline src/test/java/wolox/training/controllers/BookControllerTest.java]: **Recommendation:**
These code lines you would use in another class to use as builder.

File: src/test/java/wolox/training/controllers/BookControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    @Test
+    void whenFindByAllWhichExist_thenBooksIsReturned() throws Exception {
+        String jsonBooks = mapper.writeValueAsString(Collections.singletonList(oneTestBook));
+        Mockito.when(mockedBookRepository.getAllBook(Mockito.any(), Mockito.any(), Mockito.any()))
+                .thenReturn(Collections.singletonList(oneTestBook));
+
+        String url = API_BOOKS;
+        mvc.perform(get(url)
```
Comment (JaimLus) [inline src/test/java/wolox/training/controllers/BookControllerTest.java]: Use API_BOOKS variable directly in get method

File: src/test/java/wolox/training/controllers/BookControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    @Test
+    public void whenCreateBook_thenBookIsReturned() throws Exception {
+        String jsonBook = mapper.writeValueAsString(oneTestBook);
+        String jsonBookCreated = mapper.writeValueAsString(oneTestBookCreated);
+        Mockito.when(mockedBookRepository.save(Mockito.any())).thenReturn(oneTestBookCreated);
+
+        String url = API_BOOKS;
+        mvc.perform(post(url)
```
Comment (JaimLus) [inline src/test/java/wolox/training/controllers/BookControllerTest.java]: Use API_BOOKS variable directly in post method

File: src/test/java/wolox/training/controllers/UserControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+
+    @Test
+    void whenFindByAllWhichExist_thenUsersIsReturned() throws Exception {
+        String jsonUsers = mapper.writeValueAsString(Collections.singletonList(oneTestUser));
+        Mockito.when(mockedUserRepository.findAll()).thenReturn(Collections.singletonList(oneTestUser));
+
+        String url = API_USERS;
+        mvc.perform(get(url)
```
Comment (JaimLus) [inline src/test/java/wolox/training/controllers/UserControllerTest.java]: Use API_USERS variable directly in get method

File: src/test/java/wolox/training/controllers/UserControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+                .andExpect(content().json(jsonUser));
+    }
+
+    @Test
+    public void whenCreateUser_thenUserIsReturned() throws Exception {
+        String jsonUser = mapper.writeValueAsString(oneTestUser);
+        String jsonUserCreated = mapper.writeValueAsString(oneTestUserCreated);
+        Mockito.when(mockedUserRepository.save(Mockito.any())).thenReturn(oneTestUserCreated);
+
+        mvc.perform(post(API_USERS)
```
Comment (JaimLus) [inline src/test/java/wolox/training/controllers/UserControllerTest.java:142]: Use API_USERS variable directly in post method

**Review by JaimLus — COMMENTED** [review]
Nice job!

**Review by JaimLus — APPROVED** [review]
Nice job!

**Review by Jimenarosello — APPROVED** [review]
🚀

**Pull request merged**